CHEF-207 Validate and improve document of Chef Backend to Automate HA migration flow (#7802)
chef/automate: components/docs-chef-io/content/automate/ha_chef_backend_to_automate_ha.md

{{< /warning >}}

This page explains the procedure to migrate the existing Standalone Chef Infra Server or Chef Backend data to the newly deployed Chef Automate HA. This migration involves two steps:

- Back up the data from an existing Chef Infra Server or Chef Backend via `knife-ec-backup`.
- Restore the backed-up data to the newly deployed Chef Automate HA environment via `knife-ec-restore`.

Take a backup using the `knife-ec-backup` utility and move the backup folder to the newly deployed Chef Server. Later, restore using the same utility. The backup migrates all the cookbooks, users, data-bags, policies, and organizations.

`knife-ec-backup` utility backups and restores the data in an Enterprise Chef Server installation, preserving the data in an intermediate, editable text format. It is similar to the knife download, uploads commands, and uses the same underlying libraries. It includes workarounds for unsupported objects by the tools and various Server API deficiencies. The goal is to improve knife download, knife upload, and the Chef Infra Server API to deprecate the tool.

{{< note >}}

- The migration procedure is tested on Chef Server versions 14+ and 15+.

> - The migration procedure is tested and is possible on the Chef Backend version above 3.1.1.

Reviewer: We can change it back to 2.1.0. We hav tested on that also now.

Reviewer 2: Let's move it back to 2.1.0 then.

    ```cmd
        hab pkg exec chef/knife-ec-backup knife tidy server report --node-threshold 60 -s <chef server URL> -u <pivotal> -k <path of pivotal>
    ```

    For Example:

    ```cmd
        hab pkg exec chef/knife-ec-backup knife tidy server report --node-threshold 60 -s https://chef.io -u pivotal -k /etc/opscode/pivotal.pem
    ```

4. Execute the below command to initiate a backup of your Chef Server data. In this command:

    - `with-user-sql` is required to handle user passwords and ensure user-specific association groups are not duplicated.
    - `--with-key-sql` is to handle cases where customers have users with multiple pem keys associated with their user or clients. The current chef-server API only dumps the default key. Sometimes, users will generate and assign additional keys to give additional users access to an account but still be able to lock them out later without removing everyone's access.

    ```cmd
        hab pkg exec chef/knife-ec-backup knife ec backup backup_$(date '+%Y%m%d%H%M%s') --webui-key /etc/opscode/webui_priv.pem -s <chef server

>     ```

Reviewer: it shpuld be <chef server url> ?

    ```bash
        hab pkg exec chef/knife-ec-backup knife tidy server clean --backup-path /path/to/an-ec-backup
    ```

5. Execute the below command to copy the backup directory to the Automate HA Chef Server.

    ```cmd
        scp -i /path/to/key backup\_$(date '+%Y%m%d%H%M%s') user@host:/home/user
    ```

    If your HA Chef Server is in a private subnet, scp backup file to bastion and then to Chef Server.

## Restore Backed Up Data to Chef Automate HA

- Execute the below command to install the habitat package for `knife-ec-backup`

    ```cmd
        hab pkg install chef/knife-ec-backup
    ```

- Execute the below command to restore the backup.

    ```cmd
        hab pkg exec chef/knife-ec-backup knife ec restore /home/centos/backup\_2021061013191623331154 -yes --concurrency 1 --webui-key /hab/svc/automate-cs-oc-erchef/data/webui\_priv.pem --purge -c /hab/pkgs/chef/chef-server-ctl/*/*/omnibus-ctl/spec/fixtures/pivotal.rb
    ```

## Steps to validate if Migration is successful

- Execute the below command from the old server where a knife is installed and from the new server where the knife is installed :

    ```cmd

>         bash infra_server_objects_count_collector.sh -S test -K Key -F Filename

Reviewer: Where to find the file, we have to mention this.

- Back up the data from an existing Chef Infra Server or Chef Backend via `knife-ec-backup`.
- Restore the backed-up data to the newly deployed Chef Automate HA environment via `knife-ec-restore`.

Take a backup using the `knife-ec-backup` utility and move the backup folder to the newly deployed Chef Server. Later, restore using the same utility. The backup migrates all the cookbooks, users, data-bags, policies, and organizations.

`knife-ec-backup` utility backups and restores the data in an Enterprise Chef Server installation, preserving the data in an intermediate, editable text format. It is similar to the knife download, uploads commands, and uses the same underlying libraries. It includes workarounds for unsupported objects by the tools and various Server API deficiencies. The goal is to improve knife download, knife upload, and the Chef Infra Server API to deprecate the tool.

{{< note >}}

- The migration procedure is tested on Chef Server versions 14+ and 15+.
- The migration procedure is tested and is possible on the Chef Backend version above 3.1.1.

{{< /note >}}

## Prerequisites

> Check the [Prerequisites](/automate/ha_aws_deployment_prerequisites/#migration) page before migrating.

Reviewer 2: We should mention for for both Onprem as well as AWS

## Backup the Existing Chef Infra Server or Chef Backend Data

> 1. Execute the below command to install Habitat:

Reviewer 2: Do we need to install tab while running `knife-ec-backup` on a Chef Server or Instance?

Reviewer 2: This doc talks about installing using `ruby gem`
https://github.com/chef/knife-ec-backup#chef-infra-server-install-recommended

Author: @kalroy We have tested migration by installing hab and then knife-ec-backup package only

    ```cmd
        curl https://raw.githubusercontent.com/habitat-sh/habitat/master/components/hab/install.sh \ | sudo bash
    ```

2. Execute the below command to install the habitat package for `knife-ec-backup`

    ```cmd
        hab pkg install chef/knife-ec-backup
    ```

3. Execute the below command to generate a knife tidy server report to examine the stale node, data, etc. In this command:

    - `pivotal` is the name of the user
    - `path of pivotal` is where the user's pem file is stored.
    - `node-threshold NUM_DAYS` is the maximum number of days since the last checking before a node is considered stale.

>     ```cmd

Reviewer 2: Opinion: the command should go above before the details of the parameters.
cc: @dishanktiwari2501

        hab pkg exec chef/knife-ec-backup knife tidy server report --node-threshold 60 -s <chef server URL> -u <pivotal> -k <path of pivotal>
    ```

    For Example:

    ```cmd
        hab pkg exec chef/knife-ec-backup knife tidy server report --node-threshold 60 -s https://chef.io -u pivotal -k /etc/opscode/pivotal.pem
    ```

4. Execute the below command to initiate a backup of your Chef Server data. In this command:

>     - `with-user-sql` is required to handle user passwords and ensure user-specific association groups are not duplicated.

Reviewer 2: ```suggestion
    - `--with-user-sql` is required to handle user passwords and ensure user-specific association groups are not duplicated.
```

    ```

    - Execute the below command to clean unused data from reports. This is an optional step

    ```bash
        hab pkg exec chef/knife-ec-backup knife tidy server clean --backup-path /path/to/an-ec-backup
    ```

5. Execute the below command to copy the backup directory to the Automate HA Chef Server.

    ```cmd
        scp -i /path/to/key backup\_$(date '+%Y%m%d%H%M%s') user@host:/home/user
    ```

    If your HA Chef Server is in a private subnet, scp backup file to bastion and then to Chef Server.

## Restore Backed Up Data to Chef Automate HA

- Execute the below command to install the habitat package for `knife-ec-backup`

    ```cmd
        hab pkg install chef/knife-ec-backup
    ```

- Execute the below command to restore the backup.

    ```cmd
        hab pkg exec chef/knife-ec-backup knife ec restore <path/to/directory/backup_file> -yes --concurrency 1 --webui-key /hab/svc/automate-cs-oc-erchef/data/webui\_priv.pem --purge -c /hab/pkgs/chef/chef-server-ctl/*/*/omnibus-ctl/spec/fixtures/pivotal.rb
    ```

## Steps to validate if Migration is successful

> - Execute the below command from the old server where a knife is installed and from the new server where the knife is installed :

Reviewer 2: Let's be more explicit and split this statements into two steps:
- Fetching for older server. 
  <commad>
- Fetching from HA
<command>
-  Check the difference

    ```cmd

>         curl https://raw.githubusercontent.com/chef/automate/main/dev/infra_server_objects_count_collector.sh -o infra_server_objects_count_collector.sh

Reviewer 2: These are two different commands

Reviewer: done

        bash infra_server_objects_count_collector.sh -S test -K Key -F Filename
    ```

- `-S` is the Chef Server URL
- `-K` is the path of pivotal.pem file
- `-F` is the path to store the output file
- Now run the below command to check the differences between the old and new data. Ideally, there should be no differences if the migration was done successfully.

    ```cmd
        diff old_server_file new_server_file
    ```

> ## In-place Migration (Chef Backend to Automate HA)

Reviewer 2: We should add a section to migrate Chef Server/Backend with S3 to Automate HA. We can talk about two sub scenarios here:
- If we want to use the same S3 bucket
- If we want to use a different S3 bucket
The section should talk about steps to configure and migrate as well the implications of these two approaches

Reviewer: We have added a section. Can you please have a look. @kalroy

    endpoint = ""
    # [Optional] Mention object_storage region if applicable
    # Eg: region = "us-west-1"
    region = ""
    ## === ===
    [automate.config]
    # admin_password = ""
    # automate load balancer fqdn IP or path
    fqdn = "chef.example.com"
    instance_count = "2"
    # teams_port = ""
    config_file = "configs/automate.toml"
    [chef_server.config]
    instance_count = "2"
    [opensearch.config]
    instance_count = "3"
    [postgresql.config]
    instance_count = "3"
    [existing_infra.config]
    automate_private_ips = ["10.0.1.0","10.0.2.0"]
    chef_server_private_ips = ["10.0.1.0","10.0.2.0"]
    opensearch_private_ips = ["10.0.3.0","10.0.4.0","10.0.5.0"]
    postgresql_private_ips = ["10.0.3.0","10.0.4.0","10.0.5.0"]
    ```

9. Deploy using the following command:

```cmd
./chef-automate deploy config.toml --airgap-bundle <airgapped bundle name>
```

10. Clean up the old packages from the chef-backend (like Elasticsearch and Postgres)

11. [Restore Backed Up Data to Chef Automate HA](/automate/ha_chef_backend_to_automate_ha/#restore-backed-up-data-to-chef-automate-ha)

> ## Using Automate HA for Chef-Backend user

Reviewer 2: This title reads confusing for me. @dishanktiwari2501

Reviewer 2: We need to add a section to modify existing nodes to use the newly deployed and migrated Chef Automate HA instance. 
@YashviJain01  @shanmugapriya-tr

Reviewer: The steps are mentioned below. DO we need any change in that @kalroy

    ```cmd
        hab pkg exec chef/knife-ec-backup knife ec backup backup_$(date '+%Y%m%d%H%M%s') --webui-key /etc/opscode/webui_priv.pem -s <chef server url>
    ```
    In this command:

    - `--with-user-sql` is required to handle user passwords and ensure user-specific association groups are not duplicated.
    - `--with-key-sql` is to handle cases where customers have users with multiple pem keys associated with their user or clients. The current chef-server API only dumps the default key. Sometimes, users will generate and assign additional keys to give additional users access to an account but still be able to lock them out later without removing everyone's access.

    For example:
    ```cmd
        hab pkg exec chef/knife-ec-backup knife ec backup backup_$(date '+%Y%m%d%H%M%s') --webui-key /etc/opscode/webui_priv.pem -s https://chef.io`.
    ```

    - Execute the below command to clean unused data from reports. This is an optional step

    ```bash
        hab pkg exec chef/knife-ec-backup knife tidy server clean --backup-path /path/to/an-ec-backup
    ```

5. Execute the below command to copy the backup directory to the Automate HA Chef Server.

    ```cmd

>         scp -i /path/to/key <backup-file> user@host:/home/user

Reviewer 3: backup file or path to backup file. Just like we are writing for key

Reviewer: done.

>         scp -i /path/to/key backup\_$(date '+%Y%m%d%H%M%s') user@host:/home/user

Reviewer 3: Should we instead write <path of your backup file> something like that rather than using backup\_$(date '+%Y%m%d%H%M%s')

Reviewer: Added >backup-file> rather than any particular date with it.

    ```

    If your HA Chef Server is in a private subnet, scp backup file to bastion and then to Chef Server.

## Adding S3 Configurations for cookbook storage

Before restoring the backup on the Automate HA Chef Server, If S3 storage is required for cookbooks. Please [Configure S3](/automate/chef_infra_external_cookbooks_in_chef_automate/)

{{< note >}}

- If using same bucket for Automate HA , new files will be uploaded for cookbooks. There will not be any impact on the old files.

> - If using the a new bucket fot Auomate HA, new files will be uplaoded for cookbooks

Reviewer 3: It has few grammatical error
1. There is a space between comma and HA
2. fot -> for
3 auomate -> automate 
4 uplaoded -> uploaded

The sentence <new files will be uploaded for cookbooks.> is not clear to me

{{< /note >}}

The migration flow has been tested on following scenaiors 

> - If the cookbooks were not stored in s3, at time of migration it can be resorted in S3.

Reviewer 3: at the time of migration ?

- If the cookbooks were stored in s3, at time of migration it can either be restored in s3 or postgres.

## Restore Backed Up Data to Chef Automate HA

> - Execute the below command to install the habitat package for `knife-ec-backup`

Reviewer 3: Execute the below command where ? I mean should we mention on which machine user should run this command

Reviewer: I think thats not required. As it can be run anywhere.

    ```cmd
        hab pkg install chef/knife-ec-backup
    ```

- Execute the below command to restore the backup.

    ```cmd
        hab pkg exec chef/knife-ec-backup knife ec restore <path/to/directory/backup_file> -yes --concurrency 1 --webui-key /hab/svc/automate-cs-oc-erchef/data/webui\_priv.pem --purge -c /hab/pkgs/chef/chef-server-ctl/*/*/omnibus-ctl/spec/fixtures/pivotal.rb
    ```

## Steps to validate if Migration is successful

- Execute the below commands from the old server where the knife is installed :

    ```cmd
        curl https://raw.githubusercontent.com/chef/automate/main/dev/infra_server_objects_count_collector.sh -o infra_server_objects_count_collector.sh

>         bash infra_server_objects_count_collector.sh -S test -K Key -F Filename

Reviewer 3: I think instead of test we can write <chef-server-url> to make it consistent like other places
Same for key. I think it should be <path to key>

- Repeat the above commands on the new server where the knife is installed 

- Now run the below command to check the differences between the old and new data. Ideally, there should be no differences if the migration was done successfully.

    ```cmd
        diff old_server_file new_server_file
    ```

## In-place Migration (Chef Backend to Automate HA)

As part of this scenario, the customer will migrate from the chef-backend (5 machines) to Automate HA in place, i.e., Automate HA will be deployed in those five machines only where Chef-backend is running. One extra bastion node will be required to manage the deployment of Automate HA on the chef backend infrastructure.

{{< note >}} In-place Migration will require downtime, so plan accordingly. A reduced performance should be expected with this. {{< /note >}}

{{< note >}}

- Set up your workstation based on the newly created Automate-HA's chef-server. It is only needed if you have set up the workstation earlier.
- This in-place migration works only when cookbooks are stored in a database. This does not support use-case, where cookbooks are stored in the filesystem.

{{< /note >}}

{{< note >}}

> - In order to validate the In-place migration, please run the validation script before starting backup and restore and then once the below steps from 1 to 11 are done successfully, run it again.  

Reviewer 3: Should we make 1 to 11 are done 
to 
1 to 11 ran

{{< /warning >}}

{{< warning >}}

- Customers using only **Standalone Chef Infra Server** or **Chef Backend** are advised to follow this migration guidance. Customers using **Chef Manage** or **Private Chef Supermarket** with Chef Backend should not migrate with this.
- Also, for the customers using a standalone Chef Infra Server, cookbooks should be in the database or S3 but not in the file system.

> - Automate HA does not support supermarket authentication with chef-server user credentials.
> - Post Migration Customer can not log in with chef-server users to Supermarket.

Reviewer 4: ```suggestion
- Automate HA does not support supermarket authentication with chef-server user credentials. Post migration to Automate HA those customer users will not be able to log in with chef-server user credentials to their Supermarket.
```

{{< /warning >}}

This page explains the procedure to migrate the existing Standalone Chef Infra Server or Chef Backend data to the newly deployed Chef Automate HA. This migration involves two steps:

> - Back up the data from an existing Chef Infra Server or Chef Backend via `knife-ec-backup`.

Reviewer 4: ```suggestion
- Backup the data from an existing Chef Infra Server or Chef Backend via `knife-ec-backup`.
```

- Restore the backed-up data to the newly deployed Chef Automate HA environment via `knife-ec-restore`.

> Take a backup using the `knife-ec-backup` utility and move the backup folder to the newly deployed Chef Server. Later, restore using the same utility. The backup migrates all the cookbooks, users, data bags, policies, and organizations. `knife-ec-backup` utility backups and restores the data in an Enterprise Chef Server installation, preserving the data in an intermediate, editable text format. It is similar to the knife download, uploads commands, and uses the same underlying libraries. It includes workarounds for unsupported objects by the tools and various Server API deficiencies. The goal is to improve knife download, knife upload, and the Chef Infra Server API to deprecate the tool.

Reviewer 4: Why is this needed ?
`It is similar to the knife download, uploads commands, and uses the same underlying libraries. It includes workarounds for unsupported objects by the tools and various Server API deficiencies. The goal is to improve knife download, knife upload, and the Chef Infra Server API to deprecate the tool.`
I feel we can remove this.
```suggestion
Take a backup using the `knife-ec-backup` utility and move the backup folder to the newly deployed Chef Server. Later, restore using the same utility. The backup migrates all the cookbooks, users, data bags, policies, and organizations. `knife-ec-backup` utility backups and restores the data in an Enterprise Chef Server installation, preserving the data in an intermediate, in editable text format.
```

{{< note >}}

- The migration procedure is tested on Chef Server versions 14+ and 15+.

> - The migration procedure is tested and is possible on the Chef Backend version above 2.1.0.

Reviewer 4: I think @kalroy  mentioned we wanted to change the min. supported version to 3.x. 
@kalroy  please confirm on this.
